### ISSUE TITLE
Had to change CURL_MAX_WRITE_SIZE to get curl_formadd(CURLFORM_PTRCONTENTS) to work on 17k data

### ISSUE THREAD

**Issue by daboul** [curl/curl#2198]
Hi. I'm trying to send a ~17k data in a multipart/formdata request, and it's failing unless I overwrite the defined value of CURL_MAX_WRITE_SIZE (16384).

I'm using libcurl 7.56.0, please find below the code to replicate. 

Basically, if I use CURLFORM_PTRCONTENTS, or any other way to pass my data that I've found here https://curl.haxx.se/libcurl/c/curl_formadd.html other than CURLFORM_FILECONTENT which is working fine, then I'm limited to the size of data I can send to the server. 

A data size of 16251 or more will fail, i.e. I'm not getting my final X (16247, 16249 work).

Only way to get rid of that is to change in libcurl source code the line defining CURL_MAX_WRITE_SIZE:

```
//#define CURL_MAX_WRITE_SIZE 16384
#define CURL_MAX_WRITE_SIZE 32768
```

I don't think it is the expected behavior, can you confirm that to me?

Is there another solution than changing this define? Should there be a bug fix? I've tried changing `curl_easy_setopt(hnd, CURLOPT_BUFFERSIZE, 2*dataUTF8.size());` with no luck.

Thank you.

```c
/****************************************************************************/
std::string wstring_to_utf8 (const std::wstring& wideString)
{
std::wstring_convert<std::codecvt_utf8<wchar_t>> myconv;
return myconv.to_bytes(wideString);
}
/****************************************************************************/
int testFormCurl2(void)
{
CURLcode ret;
CURL *hnd;
struct curl_httppost *post1;
struct curl_httppost *postend;

post1 = NULL;
postend = NULL;

std::wstring dataUTF16 = _T(""); //UTF-16 NO BOM
for (int i = 0; i < 8125; i++)
{
	dataUTF16 += _T("£");
}
dataUTF16 += _T("X");

std::string dataUTF8 = wstring_to_utf8(dataUTF16); //size will be 16251 bytes

curl_formadd(&post1, &postend,
	CURLFORM_COPYNAME, "request",
	CURLFORM_PTRCONTENTS, dataUTF8.c_str(),
	CURLFORM_CONTENTTYPE, "text/plain;charset=UTF-8",
	CURLFORM_END);

// working fine from a file
//curl_formadd(&post1, &postend,
//            CURLFORM_COPYNAME, "request",
//            CURLFORM_FILECONTENT, "sent3.txt",
//            CURLFORM_CONTENTTYPE, "text/plain;charset=UTF-8",
//            CURLFORM_END);

hnd = curl_easy_init();
//curl_easy_setopt(hnd, CURLOPT_BUFFERSIZE, 2*dataUTF8.size()); // not fixing my case
curl_easy_setopt(hnd, CURLOPT_URL, "http://myurl//Rqt");
curl_easy_setopt(hnd, CURLOPT_NOPROGRESS, 1L);
curl_easy_setopt(hnd, CURLOPT_HTTPPOST, post1);
curl_easy_setopt(hnd, CURLOPT_USERAGENT, "curl/7.47.1");
curl_easy_setopt(hnd, CURLOPT_MAXREDIRS, 50L);
curl_easy_setopt(hnd, CURLOPT_SSH_KNOWNHOSTS, "C:\\Users\\xxx\\AppData\\Roaming/_ssh/known_hosts");
curl_easy_setopt(hnd, CURLOPT_COOKIEFILE, "NetCliCookie.txt");
curl_easy_setopt(hnd, CURLOPT_CUSTOMREQUEST, "POST");
curl_easy_setopt(hnd, CURLOPT_VERBOSE, 1L);
curl_easy_setopt(hnd, CURLOPT_TCP_KEEPALIVE, 1L);

ret = curl_easy_perform(hnd);

curl_easy_cleanup(hnd);
hnd = NULL;
curl_formfree(post1);
post1 = NULL;

return (int)ret;
}
```

**Comment by jay** [comment]
You should not have to change CURL_MAX_WRITE_SIZE. What do you mean will fail? Are you sure the content is encoded properly? This will not necessarily work the way you think it will:
~~~c
std::wstring dataUTF16 = _T(""); //UTF-16 NO BOM
for (int i = 0; i < 8125; i++)
{
	dataUTF16 += _T("£");
}
dataUTF16 += _T("X");
~~~
If `_UNICODE` is not defined then that will not work at all so I will assume your project has character set `Use Unicode Character Set` (just means Visual Studio defines `UNICODE` and `_UNICODE`). Since you are using `_UNICODE` then `_T()` prepends an `L` to widen the string so `dataUTF16 += _T("£");` becomes `dataUTF16 += L"£";`. Then it depends on what `£` is encoded as in the source file and whether the compiler does any encoding conversion.

(Note £ is the pound sign, unicode character U+00A3.)

There are at least 3 possible encodings I can think of in the file `A3` (ANSI) or `A3 00` (UTF-16) or `C2 A3` (UTF-8). For the first one the compiler would convert it to `A3 00`. For the last one it would convert it to `C2 00 A3 00` which is not correct.

I suspect the last one is what's happening. It's read in as this:
![utf8_c2a3_as_wchar_t](https://user-images.githubusercontent.com/965580/34421983-ad6488b4-ebe0-11e7-9d7f-b04d3de2e844.PNG)

Then when that's converted from UTF-16 back to UTF-8 it becomes this:
![utf16_c200a300_converted_back_to_utf8](https://user-images.githubusercontent.com/965580/34426180-a89cbe52-ec01-11e7-8484-7a3e5fb5098a.PNG)

My guess would be check your encodings and examine the memory and size to determine if what you think you're submitting is actually what's going to be submitted. You may want to read [this](https://github.com/jay/compiler_string_test) as well.

Also get rid of `curl_easy_setopt(hnd, CURLOPT_CUSTOMREQUEST, "POST");`, you can read why [here](https://daniel.haxx.se/blog/2015/09/11/unnecessary-use-of-curl-x/) (Note -X is same as CUSTOMREQUEST)

**Comment by jay** [comment]
I tried 7.56 this evening and was able to reproduce this. It was fixed in 5f9e2ca which is in 7.56.1. I suggest updating to the curl 7.57. If you still have problems after that then investigate the encoding.

**Comment by daboul** [comment]
Thank you. I'm pretty sure there is no encoding issue here, I've monitored everything closely with a memory inspector, I should have given a simpler example with ASCII char, sorry about that. I'll give it a try today.
